src/codec.rs

```diff
@@ -637,7 +641,87 @@ pub(crate) fn encode_slice_no_len<T: Encode, W: Output + ?Sized>(slice: &[T], de
 	}
 }
 
-/// Decode the slice (without prepended the len).
+/// Decode the array.
+///
+/// This is equivalent to decoding all the element one by one, but it is optimized for some types.
+#[inline]
+pub(crate) fn decode_array<I: Input, T: Decode, const N: usize>(input: &mut I) -> Result<[T; N], Error> {
+	#[inline]
+	fn general_array_decode<I: Input, T: Decode, const N: usize>(input: &mut I) -> Result<[T; N], Error> {
+		let mut uninit = <MaybeUninit<[T; N]>>::uninit();
+		// The following line coerces the pointer to the array to a pointer
+		// to the first array element which is equivalent.
+		let mut ptr = uninit.as_mut_ptr() as *mut T;
+		for _ in 0..N {
+			let decoded = T::decode(input)?;
+			// SAFETY: We do not read uninitialized array contents
+			//		 while initializing them.
+			unsafe {
+				ptr::write(ptr, decoded);
+			}
+			// SAFETY: Point to the next element after every iteration.
+			//		 We do this N times therefore this is safe.
+			ptr = unsafe { ptr.add(1) };
+		}
+		// SAFETY: All array elements have been initialized above.
+		let init = unsafe { uninit.assume_init() };
+		Ok(init)
+	}
+
+	// Description for the code below.
+	// It is not possible to transmute `[u8; N]` into `[T; N]` due to this issue:
+	// 		https://github.com/rust-lang/rust/issues/61956
+	//
+	// Workaround: Transmute `&[u8; N]` into `&[T; N]` and interpret that reference as value.
+	// ```
+	// let mut array: ManuallyDrop<[u8; N]> = ManuallyDrop::new([0; N]);
+	// let ref_typed: &[T; N] = unsafe { mem::transmute(&array) };
+	// let typed: [T; N] = unsafe { ptr::read(ref_typed) };
+	// Here `array` and `typed` points on the same memory.
+	// Function returns `typed` -> it is not dropped, but `array` will be dropped.
+	// To avoid that `array` under the `ManuallyDrop`.
+	// ```
+	macro_rules! decode {
+		( u8 ) => {{
+			let mut array: ManuallyDrop<[u8; N]> = ManuallyDrop::new([0; N]);
+			input.read(&mut array[..])?;
```

Comment:
This is a memory leak if `read` returns an error.

Reply:
Good catch!=) @Robbepop I returned back the usage of `forget`